processing/src/main/java/io/druid/query/extraction/NamespacedExtraction.java

```diff
@@ -0,0 +1,159 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.extraction;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.inject.name.Named;
+import com.metamx.common.StringUtils;
+
+import javax.annotation.Nullable;
+import javax.validation.constraints.NotNull;
+import java.nio.ByteBuffer;
+
+/**
+ * Namespaced extraction is a special case of DimExtractionFn where the actual extractor is pulled from a map of known implementations.
+ * In the event that an unknown namespace is passed, a simple reflective function is returned instead.
+ */
+public class NamespacedExtraction extends DimExtractionFn
+{
+  private static final byte CACHE_TYPE_ID = 0x05;
+  private static final Function<String, String> NOOP_EXTRACTOR = new Function<String, String>()
+  {
+    @Nullable
+    @Override
+    public String apply(@Nullable String input)
+    {
+      // DimExtractionFn says do not return Empty strings
+      return Strings.isNullOrEmpty(input) ? null : input;
+    }
+  };
+
+  private final String namespace;
+  private final boolean retainMissingValue;
+  private final String replaceMissingValueWith;
+  private final Function<String, String> extractionFunction;
+  private final ExtractionType extractionType;
+
+  @JsonCreator
+  public NamespacedExtraction(
+      @Nullable @JacksonInject @Named("dimExtractionNamespace")
+      final Function<String, Function<String, String>> namespaces,
+      @NotNull @JsonProperty("namespace")
+      final String namespace,
+      @Nullable @JsonProperty("retainMissingValue")
+      final Boolean retainMissingValue,
+      @Nullable @JsonProperty("replaceMissingValueWith")
+      final String replaceMissingValueWith,
+      @Nullable @JsonProperty("uniqueRenames")
+      final Boolean uniqueRenames
+  )
+  {
+    Preconditions.checkNotNull(namespace);
+    this.namespace = namespace;
+    this.retainMissingValue = retainMissingValue == null ? false : retainMissingValue;
+    this.replaceMissingValueWith = Strings.isNullOrEmpty(replaceMissingValueWith) ? null : replaceMissingValueWith;
+    Preconditions.checkArgument(!(this.retainMissingValue && !Strings.isNullOrEmpty(this.replaceMissingValueWith)), "Cannot specify a [replaceMissingValueWith] and set [retainMissingValue] to true");
+
+    if (namespaces == null) {
```

Comment:
This seems to be saying that it is acceptable to use a namespace extractor when there aren't any namespaces to extract.  That seems... weird.